File: packages/silo-client/src/query/sort-term.ts
/** Ascending unless `descending()` was called. */
export type SortDirection = "asc" | "desc";

/** One sort key: a path plus a direction. Stringifies to what the `sort`
 *  query parameter wants — the path alone when ascending, `-`-prefixed when
 *  descending. */
export class SortTerm {
  constructor(
    private readonly path: string,
    private readonly direction: SortDirection = "asc",
  ) {}

  ascending(): SortTerm {
    return new SortTerm(this.path, "asc");
  }

  descending(): SortTerm {
    return new SortTerm(this.path, "desc");
  }

  toString(): string {
    return this.direction === "desc" ? `-${this.path}` : this.path;
  }
}

File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/query/SortTerm.java
package in.org.quicko.silo.client.query;

/**
 * One sort key: a path plus a direction. Renders to what the {@code sort} query
 * parameter wants — the path alone when ascending, prefixed with {@code -} when
 * descending.
 */
public record SortTerm(String path, SortDirection direction) {

  public SortTerm(String path) {
    this(path, SortDirection.ASCENDING);
  }

  public SortTerm ascending() {
    return new SortTerm(path, SortDirection.ASCENDING);
  }

  public SortTerm descending() {
    return new SortTerm(path, SortDirection.DESCENDING);
  }

  @Override
  public String toString() {
    return direction == SortDirection.DESCENDING ? "-" + path : path;
  }
}
